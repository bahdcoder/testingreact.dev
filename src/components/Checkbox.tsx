import styled from 'styled-components'
import React, { FunctionComponent, ChangeEvent, useState, useCallback } from 'react'

interface CheckboxPropsInterface {
  label: string
  id: string
  checked?: boolean
  background?: string
  checkMarkBackground?: string
  borderColor?: string
  onChange: (event: ChangeEvent) => void
}

const Checkbox: FunctionComponent<CheckboxPropsInterface> = ({
  label,
  id,
  onChange,
  borderColor,
  checked = false,
  background = '#000',
  checkMarkBackground = '#fff',
}) => {
  const [isChecked, setIsChecked] = useState(checked)
  const callback = useCallback((e) => {
    setIsChecked(!isChecked)
    onChange(e)
  }, [])

  return <Wrapper data-testid="Checkbox">
    <Input checked={isChecked} onChange={callback} type="checkbox" id={id} />
    <Label
      htmlFor={id}
      background={background}
      borderColor={borderColor}
      checkMarkBackground={checkMarkBackground}
    >
      <LabelText>{label}</LabelText>
    </Label>
  </Wrapper>
}

const Wrapper = styled.div``

const LabelText = styled.span`
  display: inline-block;
  margin-left: 16px;
  margin-top: 3px;
  font-size: 14px;
  margin: 3px 0px 16px 16px;
  color: rgb(155, 155, 155);
  transition: 0.25s ease-in-out;
  text-transform: capitalize;
  font-weight: bolder;
`

const Label = styled.label<{
  background: string
  borderColor?: string
  checkMarkBackground: string
}>`
  position: relative;
  display: inline-block;
  padding-left: 20px;
  cursor: pointer;
  width: 100%;

  &::before {
    content: '';
    width: 20px;
    height: 20px;
    border-radius: 3px;

    position: absolute;
    top: 0px;
    left: 0px;

    display: inline-block;
    background: ${(props) => props.background};
    box-shadow: none;
    border: 1px solid
      ${(props) => (props.borderColor ? props.borderColor : props.background)};
  }
  &::after {
    content: '';
    width: 11px;
    height: 5px;

    position: absolute;
    left: 4px;
    top: 5px;

    display: inline-block;
    border-left: 2px solid ${(props) => props.checkMarkBackground};
    border-bottom: 2px solid ${(props) => props.checkMarkBackground};
    transform: rotate(-45deg);
    transition: 0.25s ease-in-out;
  }
`

const Input = styled.input`
  opacity: 0;
  position: absolute;

  & + ${Label}::after {
    opacity: 0;
  }

  &:checked + ${Label}::after {
    opacity: 1;
  }

  &:checked + ${Label} > span {
    color: rgb(0, 0, 0);
  }

  &:focus + ${Label}::before {
    outline: rgb(59, 153, 252) auto 5px;
  }
`

export default Checkbox
