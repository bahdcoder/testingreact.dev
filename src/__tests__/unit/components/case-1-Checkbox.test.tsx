import React from 'react'
import { axe } from 'jest-axe'
import { render, fireEvent, cleanup, screen, getByRole } from '@testing-library/react'
import { check } from 'prettier';

import Checkbox from '../../../components/Checkbox'

afterEach(cleanup);

const defaultProps = {
  label: 'checkbox',
  id: 'id-1',
  checked: false,
  background: '#000000',
  checkMarkBackground: '#666666',
  borderColor: '#666666',
  onChange: jest.fn(),
};

/**
 * This checkbox component renders a checkbox with a label.
 * Since we customized the default checkbox, we want to
 * make sure it still works as a regular checkbox
 * should.
 */
describe('The <Checkbox /> component', () => {
  const setupCheckbox = (props = defaultProps) =>  render(<Checkbox {...props} />);

  it('❌ Should render the label and checkbox the user will see', () => {
    const { getByTestId, debug, getByText, asFragment } = setupCheckbox();

    const checkbox = getByTestId('Checkbox');
    const label = getByText(defaultProps.label);

    // debug(label);
    // debug(checkbox);

    expect(label).toHaveTextContent(/checkbox/i)
    expect(checkbox).toBeInTheDocument();
    // or to check the snapshot
    expect(asFragment()).toMatchSnapshot();
  })

  it('❌ Should make the checkbox accessible by setting the id and htmlFor attributes on label and checkbox', () => {
  const { getByTestId, debug, getByRole, getByLabelText } =  setupCheckbox();

    const checkbox = getByTestId('Checkbox');
    const label = checkbox.querySelector('label');
    const input = getByRole('checkbox');

    // debug(label);
    // debug(checkbox);
    // debug(input);

    expect(label).toHaveAttribute('for', 'id-1');
    expect(input).toHaveAttribute('id', 'id-1');
    // or 
    expect(getByLabelText(defaultProps.label)).toBeInTheDocument();
  })

  it('❌ Should call the onChange handler when it is provided', () => {
    const { getByLabelText, debug } = setupCheckbox();

    const checkbox = getByLabelText(defaultProps.label)

    fireEvent.click(checkbox);

    // debug(checkbox);

    expect(defaultProps.onChange).toHaveBeenCalled();
  })

  it('❌ Should change state correctly when clicked (checked and unchecked)', () => {
    const { getByTestId, debug, getByRole } = setupCheckbox({...defaultProps, checked: true});

    const checkbox = getByTestId('Checkbox');
    const input = getByRole('checkbox');

    expect(input).toBeChecked();

    // fireEvent.click(input);

    // debug(input);
    // expect(input).not.toBeChecked();
  })

  it('❌ should not fail any accessibility tests', async () => {
    const { container } = setupCheckbox()

    expect(await axe(container)).toHaveNoViolations();
  })
})
