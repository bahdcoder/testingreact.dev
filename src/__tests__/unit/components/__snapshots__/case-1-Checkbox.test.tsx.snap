// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The <Checkbox /> component ‚ùå Should render the label and checkbox the user will see 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm StDqN"
    data-testid="Checkbox"
  >
    <input
      class="sc-AxhCb iqvdaa"
      id="id-1"
      type="checkbox"
    />
    <label
      class="sc-AxiKw eXpEm"
      for="id-1"
    >
      <span
        class="sc-AxirZ bZbnmP"
      >
        checkbox
      </span>
    </label>
  </div>
</DocumentFragment>
`;
